decisions/0001-bundle-granularity.rst

```diff
@@ -0,0 +1,199 @@
+================
+Bundle Granularity
+================
+
+-------
+Context
+-------
+
+Blockstore provides Bundles, which can be treated as file directories with
+versioned content. This provides a lot of flexibility, but we still need to
+determine what sets of conventions (and what extra functionality) we require to
+model our courses with OLX data.
+
+One late change to our overall philosophical outlook was to radically separate
+the human readable format used in import/export from the machine-readable format
+that provides organization and reuse capability in Blockstore itself. This
+implies shifting more of the higher level OLX -> Blockstore mapping knowledge
+to be outside of Blockstore itself, and gives us opportunities to make some
+simplifications.
+
+---------------------------------------------------
+Proposal: OLX Bundles with Flat Directory Structure
+---------------------------------------------------
+
+In this proposal, a Bundle for any chunk of OLX content has top-level folders
+for every definition in the Bundle, including both leaf and container nodes. A
+problem in a problem bank may have just one such directory for the problem
+itself, or a few directories representing a Unit and multiple leaf nodes. A
+course may potentially have thousands of directories.
+
+The major focus of this approach is operational simplicity at scale. It is
+generally less featureful than other proposals and shifts more of the work to
+OLX-aware clients.
+
+An important mental shift is that this walks back from some of the file system
+analogies of the original proposal. We still have files in Bundles, and XBlock
+content is still stored in directories with their associated static assets. But
+when we make data references across XBlocks, it's by identifiers and conventions
+and not by relative file paths. It's the job of clients to transform that into
+a more friendly file-path based format for the human readable format.
+
+XBlock Directories
+==================
+
+XBlock Directories are a client convention and not a core Blockstore primitive.
+Each directory represents a single XBlock and its associated assets. The naming
+scheme is::
+
+    /{definition-key}/
+                     {tag-type}.xml
+                     static/
+
+The ``static`` directory is optional. For a concrete example::
+
+    /problem+blockstore_granularity/
```

Comment:
Sure. In terms of Nit 1, I have no strong opinion on what the ID formatting should be.

On Nit 2: I'd prefer the slug was unique as well, but I'm not sure that's compatible with existing content. I mean, in practice it will be compatible for anything Studio authored, but no necessarily custom edited stuff.

Reply:
If you want to move towards unique slugs, that would be OK. I presume the "client" would do some validation?

Reply:
Hmm. The later is an interesting question. Is there going to be any validation at the Blockstore level?

Reply:
Actually, thinking on this more, leaving the type in there might make things easier for the client to get a sense of the Bundle as a whole with fewer reads...

@bradenmacdonald: What are you using in the newer opaque key types again?